drift.py

```diff
@@ -1,140 +1,77 @@
-        # Iterate through all of the dependencies in the new spec.
-        for i in range(len(concrete_spec)):
-            for name, dep in concrete_spec[i].items():
-                # Check if a dependency has been added.
-                if name not in cache_map:
-                    tags += ["dep-added"]
-                    continue
-                # Check if a dependency version is different.
-                elif dep["full_hash"] != cache_map[name]["full_hash"] and name != spec:
-                    tags += ["dep-updated"]
-                # Remove dependency from known dependencies to track if we don't hit any
-                # dependencies in the new version that use to be in the old spec.
-                cache_map.pop(name)
-        # If there are dependencies that we didn't hit in the new version mark
-        # the commit has having lost dependencies.
-        if len(cache_map) > 0:
-            tags += ["dep-deleted"]
-        return tags
-
-    def tag_variants(self, concrete_spec, spec, name):
-        tags = []
-        if "parameters" in concrete_spec[0][name]:
-            # Check for variants added or modified by comparing new spec to old.
-            for param, value in concrete_spec[0][name]["parameters"].items():
-                if param not in self.last[spec][0][name]["parameters"]:
-                    if param == "patches":
-                        tags += ["patches-added"]
-                    else:
-                        tags += ["variant-added"]
-                elif value != self.last[spec][0][name]["parameters"][param]:
-                    if param == "patches":
-                        tags += ["patches-modified"]
-                    else:
-                        tags += ["variant-modified"]
-            # Compare old to new to see if any variants were removed.
-            for param, value in self.last[spec][0][name]["parameters"].items():
-                if param not in concrete_spec[0][name]["parameters"]:
-                    if param == "patches":
-                        tags += ["patches-removed"]
-                    else:
-                        tags += ["variant-removed"]
-            return tags
+        for abstract_spec in self.specs:
+            # Separate out spec version and name.
+            verToken = abstract_spec.find("@")
+            if verToken < 0: verToken = len(abstract_spec)
+            out = run(f"spack versions --safe-only {abstract_spec}")
+            # Don't attempt to concretize if the version doesn't yet exist.
+            if abstract_spec[verToken+1:] in out.stdout:
+                out = run(f"spack spec --yaml {abstract_spec}")
+                # If concretization successful check the resulting concrete specs.
+                if out.returncode == SUCCESS:
+                    concrete_spec = spack.spec.Spec().from_yaml(out.stdout)
+                    if abstract_spec in self.last and self.last[abstract_spec] != concrete_spec:
```

Comment:
What is the logic here? What is self.last?

Reply:
`self.last` is that dictionary defined above. Basically I'm just storing the previous version of the spec we concretized on that last inflection point to compare the two to see when they differ.

Reply:
So here I'm saying "if we've concretized anything before and that spec differs from this one do...."

Reply:
So the idea is you could have two identical strings (package@x and package@x) and get a different result?

Reply:
Right! Because this is run on every commit that same abstract spec `package@x` will give you the different concrete specs over time we're marking as inflection points.

Reply:
ahh gotcha. And concrete_spec is going to be a string so we can easily compare. Okay I get this bit now.